dre-service/src/main/java/io/cdap/re/RulesDB.java

```diff
@@ -147,376 +199,399 @@ public void updateRule(String id, RuleRequest rule) throws RuleNotFoundException
-    List<String> rules = convertRulesToSet(row.getString(RULES));
+    StructuredRow rulebookRow = optionalRulebookStructuredRow.get();
+    List<String> rules = convertRulesToSet(rulebookRow.getString(RULES_COL));
     rules.remove(ruleId);
 
-    byte[][] columns = new byte[][] {
-      UPDATED, VERSION, RULES
-    };
-
-    version = version + 1;
-    byte[][] values = new byte[][] {
-      Bytes.toBytes(getCurrentTime()),
-      Bytes.toBytes(version),
-      Bytes.toBytes(Joiner.on(",").join(rules))
-    };
-    rulebook.put(toKey(rulebookId), columns, values);
+    Collection<Field<?>> rulebookFields = new ArrayList<>();
+    rulebookFields.add(Fields.stringField(ID_COL, rulebookId));
+    rulebookFields.add(Fields.longField(UPDATED_COL, getCurrentTime()));
+    rulebookFields.add(Fields.longField(VERSION_COL, rulebookRow.getLong(VERSION_COL) + 1));
+    rulebookFields.add(Fields.stringField(RULES_COL, GSON.toJson(Joiner.on(",").join(rules))));
+
+    rulebookTable.upsert(rulebookFields);
   }
 
-  public void cloneRulebook(String rulebookId) throws RuleAlreadyExistsException,
-    RulebookNotFoundException, RuleNotFoundException {
-    Row row = rulebook.get(toKey(rulebookId));
-    if (row.isEmpty()) {
-      throw new RulebookNotFoundException(
-        String.format("Rulebook '%s' not found.", rulebookId)
-      );
-    }
+  void deleteRulebook(String rulebookId) throws RulebookNotFoundException, IOException {
+    Collection<Field<?>> keyFields = new ArrayList<>();
+    keyFields.add(Fields.stringField(ID_COL, rulebookId));
+
+    Optional<StructuredRow> optionalStructuredRow = rulebookTable.read(keyFields);
 
-    int count = 1;
-    String clonedRulebookId = String.format("%s_%d", rulebookId, count);
-    while(!rulebook.get(toKey(clonedRulebookId)).isEmpty()) {
-      count++;
-      clonedRulebookId = String.format("%s_%d", rulebookId, count);
+    if (!optionalStructuredRow.isPresent()) {
+      throw new RulebookNotFoundException(String.format("Rulebook '%s' not found.", rulebookId));
     }
 
-    byte[][] columns = new byte[][] {
-      ID, DESCRIPTION, CREATED, UPDATED, USER, SOURCE, VERSION, RULES
-    };
-
-    byte[][] values = new byte[][] {
-      Bytes.toBytes(clonedRulebookId),
-      row.get(DESCRIPTION),
-      Bytes.toBytes(getCurrentTime()),
-      Bytes.toBytes(getCurrentTime()),
-      row.get(USER),
-      row.get(SOURCE),
-      Bytes.toBytes(1L),
-      row.get(RULES)
-    };
-
-    rulebook.put(toKey(clonedRulebookId), columns, values);
+    rulebookTable.delete(keyFields);
   }
 
-  public void deleteRulebook(String rulebookId) throws RulebookNotFoundException {
-    Row row = rulebook.get(toKey(rulebookId));
-    if (row.isEmpty()) {
-      throw new RulebookNotFoundException(
-        String.format("Rulebook '%s' not found.", rulebookId)
-      );
-    }
-    rulebook.delete(toKey(rulebookId));
-  }
+  JsonArray getRulebookRules(String rulebookId) throws RulebookNotFoundException, IOException {
```

Comment:
this is also a bad API, it should be returning a `List<Rule>`. Is it a lot of work to clean this up? (in a future PR)

Reply:
Not complicated, but will change it in a subsequent PR.